switch ADO.NET driver to bgrainger/MySqlConnector

caleblloyd opened the pull request:
This PR implements the change of the ADO.NET driver from Pomelo.Data.MySql to MySqlConnector.  It should be completely backwards compatible.

MySqlConnector is a true MIT-licensed ADO.NET implementation with focus on fast Async operation.  It should perform much better than Pomelo.Data.MySql based on the Oracle Driver.

The only questionable change in my mind was that I removed Storage.Internal.MySqlTypeMapping and am now relying on the default implementation.  @Kagamine if this was incorrect, could try implementing MySqlTypeMapping with the new ADO.NET driver?

caleblloyd commented:
Blocked on https://github.com/bgrainger/MySqlConnector/issues/42

bgrainger commented on src/Pomelo.EntityFrameworkCore.MySql/Storage/Internal/MySqlDatabaseCreator.cs:
```diff
@@ -232,10 +233,10 @@ private IReadOnlyList<MigrationCommand> CreateDropCommands()
         // Clear connection pool for the database connection since after the 'create database' call, a previously
         // invalid connection may now be valid.
-        private void ClearPool() => MySqlConnection.ClearPool((MySqlConnection)_connection.DbConnection);
+        private void ClearPool() => _connection?.Close();
```
MySqlConnector needs to implement the API you need here: https://github.com/bgrainger/MySqlConnector/issues/46

bgrainger commented on test/Pomelo.EntityFrameworkCore.MySql.Tests/MysqlRelationalConnectionTest.cs:
```diff
@@ -30,7 +26,7 @@ public void Can_create_master_connection_string()
                 using (var master = connection.CreateMasterConnection())
                 {
-                    Assert.Equal(@"server=localhost;port=3306;database=mysql;user id=root;password=Password12!;allowuservariables=True;pooling=False", master.ConnectionString);
+                    Assert.Equal(@"Server=localhost;Port=3306;Database=mysql;User Id=root;Password=Password12!;AllowUserVariables=True;Pooling=False", master.ConnectionString);
```
I assume this wasn't strictly necessary to use MySqlConnector (right?), but was just a cosmetic change?

caleblloyd replied:
The test framework was case sensitive. I imagine it works fine either way in MySqlConnector

Kagamine replied:
We doesn't require AllowUserVariables now. In other words, we can remove AllowUserVariables in this version.

bgrainger commented on src/Pomelo.EntityFrameworkCore.MySql/Update/Internal/MySqlBatchExecutor.cs:
```diff
@@ -4,120 +4,84 @@
+      CancellationToken cancellationToken = default(CancellationToken))
+    {
+      var rowsAffected = 0;
+      await connection.OpenAsync(cancellationToken);
```
Use `.ConfigureAwait(false)` in async library code.

See http://stackoverflow.com/a/13489639.

caleblloyd replied:
Thanks for the tip.  Next commit will implement this library-wide on awaited calls

bgrainger commented on src/Pomelo.EntityFrameworkCore.MySql/Update/Internal/MySqlBatchExecutor.cs:
```diff
@@ -4,120 +4,84 @@
-                    }
-                    rowsAffected += commandbatch.ModificationCommands.Count;
-                }
+          await commandbatch.ExecuteAsync(connection, cancellationToken);
```
`.ConfigureAwait(false)`

bgrainger commented on src/Pomelo.EntityFrameworkCore.MySql/Utilities/Json/JsonObject`1.cs:
```diff
@@ -0,0 +1,212 @@
+// ReSharper disable once CheckNamespace
+namespace System
+{
+  public class JsonObject<T> : IEquatable<JsonObject<T>>, IEquatable<JsonObject>, IEquatable<string>
```
Feels odd to put this type in the `System` namespace. Does it need to be there?

caleblloyd replied:
I agree, but it's to maintain BC with Pomelo.Data.MySQL (which this package presently imports)

bgrainger replied:
Makes sense.

Kagamine replied:
I suppose this should be pull into a new library.

bgrainger commented on src/Pomelo.EntityFrameworkCore.MySql/Utilities/Json/JsonObject`1.cs:
```diff
@@ -0,0 +1,212 @@
+
+    public override int GetHashCode()
+    {
+      return Json.GetHashCode();
```
Needs `null` check?

caleblloyd commented:
Code review fixes are implemented.  Package preview available as a drop-in replacement:

https://www.nuget.org/packages/caleblloyd-Pomelo.EntityFrameworkCore.MySql/1.0.1-prerelease-20160908

Kagamine commented:
I will have a try today

Kagamine commented:
@caleblloyd @bgrainger My test is blocked by bgrainger/MySqlConnector#47

Kagamine commented:
@caleblloyd It seems that MySqlConnector limited the GUID type must be `CHAR(36)`

bgrainger commented:
> My test has been blocked by bgrainger/MySqlConnector#47

Is the source code for this test available? (I checked quickly, but didn't see any test code in this repo that uses GUIDs.)

Kagamine commented:
@bgrainger Pomelo.EntityFrameworkCore.MySql has no functional tests yet, we still test the functions by using the repository [YuukoBlog](https://github.com/Kagamine/YuukoBlog-NETCore-MySql).

bgrainger commented:
@Kagamine please try [0.1.0-alpha18](https://www.nuget.org/packages/MySqlConnector/0.1.0-alpha18) and see if the improved `GetGuid` code fixes the problem you were seeing.

caleblloyd commented on src/Pomelo.EntityFrameworkCore.MySql/Update/Internal/MySqlBatchExecutor.cs:
```diff
@@ -4,120 +4,84 @@
-                }
+        foreach (var commandbatch in commandBatches)
+        {
+            commandbatch.Execute(connection);
```
need to reimplement locking here, so that one connection may only execute one statement at a time

Kagamine replied:
Yep

caleblloyd replied:
Requests are now synchronized here: https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/pull/44/files#diff-8db95c44c7794e2a6a7d91852db8a177R56

caleblloyd commented:
Added support for reading JSON;  Added synchronization of requests in same connection.  Blocked on https://github.com/bgrainger/MySqlConnector/pull/50

Kagamine commented:
@caleblloyd Should we pull the JsonObject into a new library?

caleblloyd commented:
@Kagamine yes, I will separate it out. Want to keep it in this repository as a separate library, or start a new repository for it? It needs to stay in the System namespace in order to maintain BC.

Kagamine commented:
Just start a new one. Pomelo.EntityFramework.Utilities?

caleblloyd commented:
@Kagamine just call it Pomelo.JsonObject and make me a maintainer. I agree, all types are in System anyways and the modern web framework needs a a Json type

bgrainger commented:
@caleblloyd I've pushed a commit that builds on this PR, but maintains the locking entirely within this library, instead of depending on MySqlConnector: https://github.com/bgrainger/Pomelo.EntityFrameworkCore.MySql/commit/7354ecf6dbf211721104943f878c07ad53c52581

bgrainger commented:
I've pushed a second commit that adds your JSON casting: https://github.com/bgrainger/Pomelo.EntityFrameworkCore.MySql/commit/0e2a88a0612f439a9414d6b997b795ec77bc5a09

Kagamine commented:
@caleblloyd You are the owner of this organization

Kagamine commented:
https://github.com/PomeloFoundation/Pomelo.JsonObject

caleblloyd commented:
Worked with @bgrainger to get synchronization and JSON working.  Separated JsonObject out to it's own library.  New preview build available on NuGet:

https://www.nuget.org/packages/caleblloyd-Pomelo.EntityFrameworkCore.MySql/1.0.1-prerelease-20160913

I think this implementation is pretty close to what we need.  I've run a few tests outside the library and they are working.  @Kagamine could you test the new development build also?

If this looks stable enough, I think we should merge it and focus on getting functional tests written so that we can start to stabilize this for a 1.1 release.

Travis CI is failing but I don't know why.  It's complaining that I am not implementing 2 abstract methods, however those methods do not appear in my framework so I cannot override them.  Maybe it's a config setting?

Kagamine commented:
project.json line 48:

```
         "System.Runtime": {
           "version": "4.0.10.0",
           "type": "build"
         }
```

the version should be 4.0.0.0

Kagamine commented:
I've tried to use 20160913 in a simple project, it works fine. I will do more tests today and feed back here.

caleblloyd commented:
> the version should be 4.0.0.0

changed the version, the warning disappeared.  Travis is still complaining about non-existent abstract methods though

Kagamine commented on src/Pomelo.EntityFrameworkCore.MySql/Storage/Internal/SynchronizedMySqlDataReader.cs:
```diff
@@ -0,0 +1,199 @@
+using MySql.Data.MySqlClient;
+
+// ReSharper disable once CheckNamespace
+namespace Microsoft.EntityFrameworkCore.Storage.Internal
```
Should the namespace be `MySql.Data.MySqlClient` is better?

Kagamine commented on test/Pomelo.EntityFrameworkCore.MySql.Tests/MysqlRelationalConnectionTest.cs:
```diff
@@ -30,7 +26,7 @@ public void Can_create_master_connection_string()
             {
                 using (var master = connection.CreateMasterConnection())
                 {
-                    Assert.Equal(@"server=localhost;port=3306;database=mysql;user id=root;password=Password12!;allowuservariables=True;pooling=False", master.ConnectionString);
```
The new `MySqlConnector` doesn't support lowercase?

Kagamine commented:
Maybe the mono version of travis is so old.

Kagamine commented:
I want to keep the version 1.0.1-prerelease-*, until the Microsoft released EF Core 1.0.1-rtm. 

I think make the version rule as same as the official packages is good for developers.

caleblloyd commented on src/Pomelo.EntityFrameworkCore.MySql/Storage/SynchronizedMySqlDataReader.cs:
```diff
@@ -1,199 +1,200 @@
+using MySql.Data.MySqlClient;
+
+// ReSharper disable once CheckNamespace
+namespace Microsoft.EntityFrameworkCore.Storage
```
@Kagamine I moved it from `Microsoft.EntityFrameworkCore.Storage.Interna`l to `Microsoft.EntityFrameworkCore.Storage`.  The Internal namespaces are typically for objects that can be changed out through Dependency Injection.  I see Microsoft using the parent namespace (`Microsoft.EntityFrameworkCore.Storage`) in this case for implementation that cannot be changed out through DI.

Let's stay out of the `MySqlConnector` namespace.  This Synchronized Reader could technically synchronize any `DbDataReader`, and has a pretty tight coupling to the lock in `Microsoft.EntityFrameworkCore.Storage.Internal.MySqlRelationalConnection`.

Kagamine commented:
:shipit:

caleblloyd commented:
Landed!  My next focus will be functional testing and performance testing against Oracle's reference implementation.  I think when people see the Async power of `MySqlConnector`, it'll be an easy choice which EF Provider to go with!

Thanks for the help on this @bgrainger @Kagamine

Kagamine commented:
Published 1.0.1-prerelease-10001

https://www.myget.org/Package/Details/pomelo?packageType=nuget&packageId=Pomelo.EntityFrameworkCore.MySql&packageVersion=1.0.1-prerelease-10001
